adam-core/src/main/scala/org/bdgenomics/adam/rdd/sets/ShuffleRegionJoin.scala

```diff
@@ -0,0 +1,408 @@
+            val intervals = partitionMapIntervals.get(f._1.pad(threshold), requireOverlap = false)
+            // for each index identified in intervals, create a record
+            intervals.map(g => ((f._1, g._2), f._2))
+          })
+        }, preservesPartitioning = true)
+        .repartitionAndSortWithinPartitions(
+          ManualRegionPartitioner(destinationPartitionMap.length))
+        .map(f => (f._1._1, f._2))
+    }
+
+    (preparedLeft, preparedRight)
+  }
+}
+
+/**
+ * Perform a ShuffleRegionJoin on a GenomicRDD.
+ *
+ * @tparam T The type of the left records.
+ * @tparam U The type of the left, a subclass of GenomicRDD.
+ * @tparam X The type of the right records.
+ * @tparam Y The type of the right, a subclass of GenomicRDD.
+ * @tparam RT The resulting type of the left after the join.
+ * @tparam RX The resulting type of the right after the join.
+ */
+private[rdd] sealed trait ShuffleRegionJoinOnGenomicRDD[T, U <: GenomicRDD[T, U], X, Y <: GenomicRDD[X, Y], RT, RX]
+    extends ShuffleRegionJoin[T, X, RT, RX] {
+
+  protected val leftRdd: GenomicRDD[T, U]
+  protected val rightRdd: GenomicRDD[X, Y]
+
+  override protected def prepare()(implicit tTag: ClassTag[T], xTag: ClassTag[X]): (RDD[(ReferenceRegion, T)], RDD[(ReferenceRegion, X)]) = {
+    prepareRdds(
+      leftRdd.flattenRddByRegions(),
+      rightRdd.flattenRddByRegions(),
+      leftRdd.partitionMap)
+  }
+}
+
+/**
+ * Perform a ShuffleRegionJoin on ReferenceRegion keyed RDDs.
+ *
+ * @tparam T The type of the left records.
+ * @tparam X The type of the right records.
+ * @tparam RT The resulting type of the left after the join.
+ * @tparam RX The resulting type of the right after the join.
+ */
+private[rdd] sealed trait ShuffleRegionJoinOnReferenceRegionKeyedRDD[T, X, RT, RX]
+    extends ShuffleRegionJoin[T, X, RT, RX] {
+
+  protected val leftRdd: RDD[(ReferenceRegion, T)]
+  protected val rightRdd: RDD[(ReferenceRegion, X)]
+
+  override protected def prepare()(implicit tTag: ClassTag[T], xTag: ClassTag[X]): (RDD[(ReferenceRegion, T)], RDD[(ReferenceRegion, X)]) = {
+    prepareRdds(leftRdd, rightRdd, PartitionMap(None))
+  }
+}
+
+/**
+ * Perform an Inner ShuffleRegionJoin.
+ *
+ * @tparam T The left side row data.
+ * @tparam X The right side row data.
+ */
+sealed trait InnerShuffleRegionJoin[T, X] extends VictimlessSetOperationBetweenCollections[T, X, T, X] {
+
+  override protected def emptyFn(left: Iterator[(ReferenceRegion, T)],
+                                 right: Iterator[(ReferenceRegion, X)]): Iterator[(T, X)] = {
+    Iterator.empty
+  }
+
+  override protected def postProcessHits(currentLeft: (ReferenceRegion, T),
+                                         iter: Iterable[(ReferenceRegion, X)]): Iterable[(T, X)] = {
+    iter.map(f => (currentLeft._2, f._2))
+  }
+}
+
+/**
+ * Perform an Inner ShuffleRegionJoin. This is publicly accessible to be
+ * compatible with legacy code.
+ */
+object InnerShuffleRegionJoin {
```

Comment:
Pinging @heuermh and @fnothaft to get your thoughts on this architecture. This was what I came up with to guarantee consistency with our previous implementation. From the user point of view, it looks exactly the same as it did before, despite looking different under the hood.